### PR TITLE
Add multicloser implementation

### PR DESCRIPTION
**Pull request by sethvargo**
Alternative approach to https://github.com/abcxyz/pkg/pull/135

### CODE REVIEW

File: multicloser/multicloser.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+//
+// It is not safe to use concurrently without locking.
+type Closer struct {
+	fns []func() error
+}
+
+// Append adds the given closer functions. It handles void and error signatures.
+// Other signatures should use an anonymous function to match an expected
+// signature.
+func Append[T Func](c *Closer, fns ...T) *Closer {
```
Comment (sethvargo) [inline multicloser/multicloser.go:42]: This works, and it's typesafe, but you can't _blend types_ in the same call.

```go
// Good
c = multicloser.Append(c, func() {}, func() {})

// Good
c = multicloser.Append(c, func() error {}, func() error {})

// BAD!
c = multicloser.Append(c, func() {}, func() error {})
```

File: multicloser/multicloser.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+package multicloser
+
+import (
+	"errors"
+)
+
+// Func is the type signature for a closing function. It accepts a function that
+// returns an error or a void function.
+type Func interface {
+	func() error | func()
```
Comment (sethvargo) [inline multicloser/multicloser.go:28]: You **cannot** make these concrete types or the type system freaks out. This is the situation I was running into the first time I tried.

**Review by yolocs — APPROVED** [review]
You beat me! Ship it!

File: multicloser/multicloser.go (hunk 1 above)
Comment (drevell) [reply multicloser/multicloser.go:42]: One possibility: add function WrapVoid function that takes a `func() void` and returns a no-op `func() error`

[the remark this answers is not in the document]
Comment (sethvargo) [reply multicloser/multicloser.go:42]: https://github.com/abcxyz/pkg/pull/137

**Pull request merged**